metrotranscom/doorway
# feat: add file upload endpoint to asset service

# Pull Request Template

## Issue Overview

This PR addresses in part #74


## Description

This PR adds an additional action to the backed `assets` service/controller to enable file uploads to pass directly through the backend, providing a single place where file uploads can be managed rather than spreading the responsibility across multiple packages.  It builds upon the work done in #80 and will be finished in a subsequent PR to update the partner site to use this backend rather than attempting to handle file uploads itself.

Some notes:

- The `UploadService` dependency is no longer required, since all it did was generate signatures and pass them to the frontend.  It should be removed in the future, but is retained in this PR due to the current version of the partner site relying upon it.
- File validation is done manually in the controller since file validators don't exist in the version of Nest.js that we are using.  This can be updated in the future when the version gets bumped.
- There is no `CloudinaryFileService`.  If we had a little more time and/or there was a chance we would actually use it, I'd add it in for the sake of completeness.  We aren't though, so I'll leave that for a future PR.

## How Can This Be Tested/Reviewed?

New unit and e2e tests have been added to validate functionality, which can be run using `yarn test` and `yarn test:e2e:local`.

There currently isn't any integration with the partner site, so testing the functionality is a bit more manual.

First, edit your .env file for the backend service to include the new values, replacing the bucket var with this: `ASSET_FS_CONFIG_s3_BUCKET=doorway-public-uploads20230505214133756500000001`.  Build and start the backend service.

Next, start up and log in to the partner site.  Once you are logged in, grab your cookie from browser dev tools.

![image](https://user-images.githubusercontent.com/23032560/236862850-0d54293a-ec56-4da5-b2b6-ad04477536b5.png)

Open up a terminal window and set it as a var named `BLOOM_COOKIE`.

![image](https://user-images.githubusercontent.com/23032560/236863677-40c98d17-d293-45fc-b260-f9d394b817e3.png)

Find a file to upload (must be a jpeg, png, or pdf) and set the path to it using `export ASSET_FILE="/path/to/file.png"`.  Then set a label for the file using `export ASSET_LABEL= "<some-label>".  It should look something like this:

![image](https://user-images.githubusercontent.com/23032560/236865060-d79073c7-749d-49e5-90fe-ec7dab31d6ae.png)

Once those are set, use `curl` to upload the file using this command: `curl -i -X POST --cookie "${BLOOM_COOKIE}" -F "file=@${ASSET_FILE}" -F "label=${ASSET_LABEL}" http://localhost:3100/assets/upload/`

You should get a response that looks something like this: 

![image](https://user-images.githubusercontent.com/23032560/236865966-a7961eac-48da-4c89-b5a3-ab631539d9ea.png)

If you query the database, you should see an entry for the asset you uploaded.

![image](https://user-images.githubusercontent.com/23032560/236867409-e90e0fa4-63ff-4a7d-bf41-e5477468b8f6.png)

And if you check the bucket in AWS, you should see your file there.

![image](https://user-images.githubusercontent.com/23032560/236867765-2f8a8db2-da9e-4abb-83f2-28672ae96878.png)


## Checklist:


## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.
